Optim foreach cleanup for Adagrad

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* __->__ #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to Adagrad optimizer + cleanup

Differential Revision: [D33767863](https://our.internmc.facebook.com/intern/diff/D33767863)

mikaylagawarecki committed:
Optim foreach cleanup for Adagrad

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/adagrad.py:
```diff
@@ -103,13 +112,135 @@ def step(self, closure=None):
+                           grads: List[Tensor],
+                           state_sums: List[Tensor],
+                           state_steps: List[int],
+                           has_sparse_grad: bool = False,
```
added an unused kwarg with a default here `has_sparse_grad` so both functions take the same inputs and the above logic looks cleaner, is this ok?

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

albanD commented on torch/optim/adagrad.py:
```diff
@@ -103,13 +116,149 @@ def step(self, closure=None):
+    if len(params) == 0:
+        return
+
+    if has_sparse_grad:
```
This should never happen given that you already check that in the main function right?

albanD commented on torch/optim/adagrad.py:
```diff
@@ -103,13 +116,149 @@ def step(self, closure=None):
+            state_steps: List[int],
+            # kwonly args with defaults are not supported by functions compiled with torchscript issue #70627
+            # setting these as kwargs for now as functional API is compiled by torch/distributed/optim
+            has_sparse_grad: bool = False,
```
Why is the default False here? (and not None)
Don't we want to check it if the user didn't provide it? Or raise an error if the user asked for foreach explicitly but didn't provide this?

albanD reviewed (commented):
Small question about has_sparse_grad. Sounds good otherwise

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

albanD commented on torch/optim/adagrad.py:
```diff
@@ -92,21 +102,157 @@ def step(self, closure=None):
+    if foreach and torch.jit.is_scripting():
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if has_sparse_grad is None:
```
We don't need to do that if we call into the non-multi-tensor impl. Given that this is not "free" we should skip that.
I think by moving this logic in the multi-tensor code is a good way to do that.

albanD reviewed (commented):
Looks good. We should just avoid some extra computation that is not needed but good to go otherwise.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"




[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767863).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adagrad"


Differential Revision: [D33767863](https://our.internmc.facebook.com/intern/diff/D33767863)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767863).
